### PR TITLE
17592  / Update confirmation message and refactoring

### PR DESCRIPTION
**Pull request by jon-shipley**
feature/17592-errors-in-confirmation-message

* update copy in the flash message after adding a new pupil
* refactoring 
   - pupil controller refactored to be more testable
   - getAddPupil postAddPupil refactored
   - ValidationError made easier to test
   - new pupilDataService.save() method
  - new DateService. createFromDayMonthYear() method that is useful when creating dates from GDS input boxes (day, month, year from user input)
  - new pupilAddService

### CODE REVIEW

File: admin/controllers/pupil.js [hunk 1]
```diff
@@ -76,15 +76,42 @@ const postAddPupil = async (req, res, next) => {
...
+    req.flash('info', '1 new pupil has been added')
   } catch (error) {
     next(error)
   }
   const pupilId = JSON.stringify([pupil._id])
   res.redirect(`/school/pupil-register/lastName/true?hl=${pupilId}`)
 }
 
-const getAddMultiplePupils = (req, res, next) => {
+controller.postAddPupil2 = async (req, res, next) => {
```
Comment (GeorgeChatzigiannis) [inline admin/controllers/pupil.js]: is that a temporary naming?

File: admin/spec/controllers/pupil.spec.js [hunk 2]
```diff
@@ -445,4 +456,68 @@ describe('pupil controller:', () => {
...
+
+      it('calls pupilAddService to add a new pupil to the database', async (done) => {
+        await controller(req, res, nextSpy)
+        expect(pupilAddServiceSpy.callCount).toBe(1)
+        done()
+      })
+
+      it('redirects to the pupil register page', async (done) => {
+        await controller(req, res, nextSpy)
+        expect(res.statusCode).toBe(302)
```
Comment (GeorgeChatzigiannis) [inline admin/spec/controllers/pupil.spec.js]: that's just a 302 code where's the redirection call?
Perhaps also check for the page title to be what you expect?

File: admin/controllers/pupil.js (hunk 1 above)
Comment (jon-shipley) [reply admin/controllers/pupil.js]: yes

File: admin/controllers/pupil.js [hunk 3]
```diff
@@ -10,81 +12,64 @@ const pupilService = require('../services/pupil.service')
 const pupilDataService = require('../services/data-access/pupil.data.service')
 const azureFileDataService = require('../services/data-access/azure-file.data.service')
 const pupilUploadService = require('../services/pupil-upload.service')
+const schoolDataService = require('../services/data-access/school.data.service')
+const pupilAddService = require('../services/pupil-add-service')
+
+const controller = {}
 
-const getAddPupil = async (req, res, next) => {
+controller.getAddPupil = async (req, res, next, error = null) => {
```
Comment (pris54) [inline admin/controllers/pupil.js]: Are we changing the naming convention? This will require some refactoring of existing controllers.

Comment (jon-shipley) [reply admin/controllers/pupil.js]: It's still the same name.  It was necessary to change the the exports for unit testing - as it was it was impossible to stub the getAddPupil from the postAddPupil.

I'd suggest this is a more testable pattern.

Comment (jon-shipley) [reply admin/controllers/pupil.js]: The call signature was updated to allow the error page to displayed, as well as the default initial page.

**Pull request merged**